src/sqlfluff/testing/rules.py

```diff
@@ -135,38 +171,22 @@ def rules__test_helper(test_case):
             configs=test_case.configs,
         )
     if test_case.fail_str:
-        res = assert_rule_fail_in_sql(
+        res, violations_before_fix = assert_rule_fail_in_sql(
             test_case.rule,
             test_case.fail_str,
             configs=test_case.configs,
             line_numbers=test_case.line_numbers,
         )
+        if test_case.violations:
+            assert_violations_before_fix(test_case, violations_before_fix)
         # If a `fixed` value is provided then check it matches
         if test_case.fix_str:
             assert res == test_case.fix_str
+            if test_case.violations_after_fix:
+                assert_violations_after_fix(test_case)
         else:
             # Check that tests without a fix_str do not apply any fixes.
             assert res == test_case.fail_str, (
                 "No fix_str was provided, but the rule modified the SQL. Where a fix "
                 "can be applied by a rule, a fix_str must be supplied in the test."
             )
-
-
-def dedent(sql: str) -> str:
-    """Cleans up the surrounding whitespace."""
-    return textwrap.dedent(sql).lstrip("\n").rstrip() + "\n"
-
-
-def fix(code: str, sql: str, configs: Optional[dict] = None) -> str:
-    """Runs fix and returns the fixed sql."""
-    return assert_rule_fail_in_sql(code, sql, configs)
-
-
-def lint(code: str, sql: str, configs: Optional[dict] = None) -> List[str]:
-    """Runs lint and returns the found violations."""
-    return [
-        v.description if isinstance(v, SQLLintError) else str(v)
-        for v in Linter(FluffConfig(configs=configs, overrides={"rules": code}))
-        .lint_string(sql)
-        .violations
-    ]
```

Comment:
Why are these three functions not needed anymore?

Reply:
I added them in https://github.com/sqlfluff/sqlfluff/pull/2681 as generic helpers for python tests, but there are no usages left after migrating the related test from python to yaml. I presume these functions are not needed in the future because the yaml framework covers the same needs.